src/gadget/ecc.rs

```diff
@@ -0,0 +1,67 @@
+//! Gadgets and chips for elliptic curve operations.
+
+use std::fmt;
+
+use crate::{
+    arithmetic::Curve,
+    circuit::{Chip, Layouter},
+    plonk::Error,
+};
+
+/// The set of circuit instructions required to use the ECC gadgets.
+pub trait EccInstructions<C: Curve>: Chip<Field = C::Base> {
+    /// Variable representing an elliptic curve point.
+    type Point: Clone + fmt::Debug;
+    /// Variable representing an element of the elliptic curve's scalar field.
+    type Scalar: Clone + fmt::Debug;
+
+    /// Performs point addition, returning `a + b`.
+    fn add(
+        layouter: &mut impl Layouter<Self>,
+        a: &Self::Point,
+        b: &Self::Point,
+    ) -> Result<Self::Point, Error>;
+
+    /// Performs point doubling, returning `[2] a`.
+    fn double(layouter: &mut impl Layouter<Self>, a: &Self::Point) -> Result<Self::Point, Error>;
+
+    /// Performs variable-base scalar multiplication, returning `[scalar] base`.
+    fn mul(
+        layouter: &mut impl Layouter<Self>,
+        base: &Self::Point,
+        scalar: &Self::Scalar,
+    ) -> Result<Self::Point, Error>;
+}
+
+/// An element of the given elliptic curve's scalar field.
+#[derive(Debug)]
+pub struct Scalar<C: Curve, EccChip: EccInstructions<C>> {
+    inner: EccChip::Scalar,
+}
+
+/// An elliptic curve point over the given curve.
+#[derive(Debug)]
+pub struct Point<C: Curve, EccChip: EccInstructions<C>> {
```

Comment:
Making `C` a type parameter of `EccInstructions` has the nice side-effect that the `Point` gadget is typed to a specific curve.